bliss/simulator/toy_simulated_dataset.py

```diff
@@ -0,0 +1,380 @@
+import dataclasses
+import math
+import warnings
+
+import torch
+from torch.utils.data import IterableDataset
+
+from bliss.catalog import FullCatalog
+
+# prevent pytorch_lightning warning for num_workers = 0 in dataloaders with IterableDataset
+warnings.filterwarnings(
+    "ignore", ".*does not have many workers which may be a bottleneck.*", UserWarning
+)
+
+
+def counts_mask(counts):
+    """Identify the first `counts[i]` elements of the ith row of the result.
+
+    Args:
+        counts: (m) tensor of counts
+
+    Returns:
+        (m,M) tensor where M = counts.max() and result[i, j] = j < counts[i]
+
+
+    Example:
+    counts = torch.tensor([2,1,1])
+    counts_mask(counts)
+    tensor([[True, True],
+            [True, False],
+            [True, False]])
+
+    """
+    # Determine the max length for the range
+    max_count = counts.max()
+
+    # Create a range tensor [0, 1, 2, ..., max_count-1]
+    range_tensor = torch.arange(max_count, device=counts.device).unsqueeze(0)
+
+    # Expand counts to match the shape for broadcasting
+    expanded_counts = counts.unsqueeze(1)
+
+    # Compute the mask
+    return range_tensor < expanded_counts
+
+
+def draw_trunc_pareto(b, c, loc, scale, size, torch_generator=None):
+    """Draw a truncated pareto random variable.
+
+    Args:
+        b: float, exponent of the power law
+        c: float, truncation of the power law
+        loc: float, location of the power law
+        scale: float, scale of the power law
+        size: tuple of ints, size of the output
+        torch_generator: torch.Generator, optional; random number generator
+
+    Returns:
+        (size) tensor of draws from the truncated pareto distribution
+
+    Specifically, let the "standardized"
+    truncated pareto distribution be the distribution with PDF
+
+    f(x) propto x^{-b-1} I(1<=x<=c)
+
+    Then we return draws from a shifted and scaled version, i.e.
+
+    X ~ f(x)
+    Y = X*scale+loc
+    """
+
+    # get generator if necessary
+    torch_generator = torch.Generator() if torch_generator is None else torch_generator
+
+    # draw from the standardized truncated pareto
+    u = torch.rand(size, generator=torch_generator, device=torch_generator.device)
+
+    # get (c^b -1) / c^b)
+    ratio = (c ** (-b) - 1) / c ** (-b)
+
+    # get unscaled draws
+    x = torch.exp(-torch.log1p(-u * ratio) / b)
+
+    # scale and shift
+    return x * scale + loc
+
+
+@dataclasses.dataclass
+class ToySimulator:
```

Comment:
Could we inherit from `pl.LightningDataModule` and `IterableDataset` here, for symmetry with `SimulatedDataset`? Ideally `ToySimulator` would be as close to a drop-in replacement for `SimulatedDataset` as possib.e

Reply:
Oh maybe this is OK, I see now it's `ToySimulatedDataset` that's intended as the drop-in replacement for `SimulatedDataset`. It does seem a bit cleaner to me to merge `ToySimulator` and `ToySimulatedDataset`, and to inherit from the same classes that `SimulatedDataset` does.